File: tests-python/web_playwright/pages/webform_page.py
from playwright.sync_api import Page, expect


class WebFormPage:

    def __init__(self, page: Page):
        self.page = page
        self.input = page.get_by_label("Text input")
        self.password = page.get_by_label("Password")
        self.textarea = page.get_by_label("Textarea")
        self.number = page.locator("//select[@name='my-select']")
        self.city = page.locator("//input[@name='my-datalist']")
        self.color = page.get_by_label("Color picker")
        self.date = page.get_by_label("Date picker")
        self.range = page.get_by_label("Example range")
        self.file = page.get_by_label("File input")
        self.button_submit = page.get_by_role("button", name="Submit")

    def set_input(self, value):
        self.input.fill(value)

    def set_password(self, value):
        self.password.fill(value)

    def set_textarea(self, value):
        self.textarea.fill(value)

    def set_number(self, value):
        self.number.select_option(index=value)

    def set_city(self, value):
        self.city.fill(value)

    def set_color(self, value):
        self.color.fill(value)

    def set_date(self, value):
        #self.date.fill(value)
        # Click elsewhere to close the calendar
        #self.page.locator("body").click()
        self.date.evaluate("(elem, val) => elem.setAttribute('value', val);", value)
        
    def set_range(self, value):
        # Two solutions
        # Solution #1: Trigger mouse events to move the range handler
        width = self.range.evaluate("elem => { return elem.getBoundingClientRect().width }")
        self.range.hover(position={'x': 0, 'y': 0})
        self.page.mouse.down()
        self.range.hover(position={'x': (width * value) / 10, 'y': 0})
        self.page.mouse.up()

        # solution #2: Set the range value directly
        self.range.fill(str(value))

    def set_file(self, value):
        self.file.set_input_files(value)

    def submit(self):
        self.button_submit.click()

File: tests-nodejs/web_playwright/pages/webform.page.ts
import { expect, type Locator, type Page } from '@playwright/test';


class WebformPage {

  page: Page;
  input: Locator;
  password: Locator;
  textarea: Locator;
  number: Locator;
  city: Locator;
  color: Locator;
  date: Locator;
  range: Locator;
  file: Locator;
  button: Locator;

  constructor(page: Page) {
    this.page = page;
    this.input = page.getByLabel("Text input");
    this.password = page.getByLabel("Password");
    this.textarea = page.getByLabel("Textarea");
    this.number = page.locator("//select[@name='my-select']");
    this.city = page.locator("//input[@name='my-datalist']");
    this.color = page.getByLabel("Color picker");
    this.date = page.getByLabel("Date picker");
    this.range = page.getByLabel("Example range");
    this.file = page.getByLabel("File input");
    this.button = page.getByRole("button", { name: "Submit" });
  }

  async set_input(value) {
    await this.input.fill(value);
  }

  async set_password(value) {
    await this.password.fill(value);
  }

  async set_textarea(value) {
    await this.textarea.fill(value);
  }

  async set_number(value) {
    await this.number.selectOption({ index: value });
  }

  async set_city(value) {
    await this.city.fill(value);
  }

  async set_color(value) {
    await this.color.fill(value);
  }

  async set_date(value) {
    await this.date.evaluate((elem, value) => elem.setAttribute("value", value), value);
  }

  async set_file(value) {
    await this.file.setInputFiles(value);
  }

  async set_range(value) {
    await this.range.fill(value.toString());
  }

  async submit() {
    await this.button.click();
  }

}

export default WebformPage;
